### PR TITLE
Fix SDK cache workflow task start to close timeout behavior

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Timeout / failure of workflow cache task should also invalidate mutable state SDK specific task queue
* Timeout / failure of workflow cache task should also insert a new workflow task schedule event

**Why?**
T = 0 SDK picks up a SDK / host specific workflow task
T = 1 SDK unable to finish workflow task, server timeout
T = 2 (actual) server retry workflow task using normal task queue without inserting workflow task schedule event
T = 2 (bugfix) server retry workflow task using normal task queue & inserting workflow task schedule event

Nothing breaks, only behavior is confusing

another example:

with bug
```bash
1  WorkflowExecutionStarted  {WorkflowType:{Name:queries_in_wf_task},
                                ParentInitiatedEventId:0,
                                TaskQueue:{Name:queries_in_wf_task_r7uhKl9r, Kind:Normal},
                                WorkflowTaskTimeout:1s, Initiator:Unspecified,
                                OriginalExecutionRunId:f83759f5-e38c-4fb0-b34e-0a3996da0d0c,
                                FirstExecutionRunId:f83759f5-e38c-4fb0-b34e-0a3996da0d0c,
                                Attempt:1, FirstWorkflowTaskBackoff:0s}
   2  WorkflowTaskScheduled     {TaskQueue:{Name:queries_in_wf_task_r7uhKl9r,
                                Kind:Normal}, StartToCloseTimeout:1s,
                                Attempt:1}
   3  WorkflowTaskStarted       {ScheduledEventId:2, Identity:integ_tester,
                                RequestId:71f7692d-6f8d-4de4-be61-291b032fb6d0}
   4  WorkflowTaskCompleted     {ScheduledEventId:2,
                                StartedEventId:3,
                                Identity:integ_tester}
   5  TimerStarted              {TimerId:timer-1,
                                StartToFireTimeout:500ms,
                                WorkflowTaskCompletedEventId:4}
   6  TimerFired                {TimerId:timer-1,
                                StartedEventId:5}
   7  WorkflowTaskScheduled     {TaskQueue:{Name:integ_tester-queries_in_wf_task_r7uhKl9r-90d9b52dfded432fadd7ed4799556c49,
                                Kind:Sticky}, StartToCloseTimeout:1s, Attempt:1}
   8  WorkflowTaskStarted       {ScheduledEventId:7, Identity:integ_tester,
                                RequestId:ab5f74b7-51be-4fd9-8152-6ffc24ab4750}
   9  WorkflowTaskTimedOut      {ScheduledEventId:7,
                                StartedEventId:8,
                                TimeoutType:StartToClose}
```

without bug
```bash
1  WorkflowExecutionStarted  {WorkflowType:{Name:queries_in_wf_task},
                                ParentInitiatedEventId:0,
                                TaskQueue:{Name:queries_in_wf_task_r7uhKl9r, Kind:Normal},
                                WorkflowTaskTimeout:1s, Initiator:Unspecified,
                                OriginalExecutionRunId:f83759f5-e38c-4fb0-b34e-0a3996da0d0c,
                                FirstExecutionRunId:f83759f5-e38c-4fb0-b34e-0a3996da0d0c,
                                Attempt:1, FirstWorkflowTaskBackoff:0s}
   2  WorkflowTaskScheduled     {TaskQueue:{Name:queries_in_wf_task_r7uhKl9r,
                                Kind:Normal}, StartToCloseTimeout:1s,
                                Attempt:1}
   3  WorkflowTaskStarted       {ScheduledEventId:2, Identity:integ_tester,
                                RequestId:71f7692d-6f8d-4de4-be61-291b032fb6d0}
   4  WorkflowTaskCompleted     {ScheduledEventId:2,
                                StartedEventId:3,
                                Identity:integ_tester}
   5  TimerStarted              {TimerId:timer-1,
                                StartToFireTimeout:500ms,
                                WorkflowTaskCompletedEventId:4}
   6  TimerFired                {TimerId:timer-1,
                                StartedEventId:5}
   7  WorkflowTaskScheduled     {TaskQueue:{Name:integ_tester-queries_in_wf_task_r7uhKl9r-90d9b52dfded432fadd7ed4799556c49,
                                Kind:Sticky}, StartToCloseTimeout:1s, Attempt:1}
   8  WorkflowTaskStarted       {ScheduledEventId:7, Identity:integ_tester,
                                RequestId:ab5f74b7-51be-4fd9-8152-6ffc24ab4750}
   9  WorkflowTaskTimedOut      {ScheduledEventId:7,
                                StartedEventId:8,
                                TimeoutType:StartToClose}
  10  WorkflowTaskScheduled     {TaskQueue:{Name:queries_in_wf_task_r7uhKl9r,
                                Kind:Normal}, StartToCloseTimeout:1s,
                                Attempt:1}
  11  WorkflowTaskStarted       {ScheduledEventId:10, Identity:integ_tester,
                                RequestId:fa0dab3f-bb28-433b-b7ed-5f53656d15d0}
  12  WorkflowTaskTimedOut      {ScheduledEventId:10,
                                StartedEventId:11,
                                TimeoutType:StartToClose}
```

**How did you test it?**
Manual

**Potential risks**
N/A

**Is hotfix candidate?**
No

### CODE REVIEW

**Review by Sushisource — APPROVED** [review]
Verified this resulted in WFT after timeout being re-scheduled onto the non-sticky queue in my test

**Comment by yiminc** [comment]
I could be wrong, but I think we intentionally skip record events for continues WFT failure to avoid growing the history events indefinitely. Those are called transient events? Those events are only recorded in history when it eventually succeed or when workflow closed.

File: service/history/workflow/workflow_task_state_machine.go [hunk 1]
```diff
@@ -562,6 +556,7 @@ func (m *workflowTaskStateMachine) FailWorkflowTask(
 	incrementAttempt bool,
 ) {
 	// clear stickiness whenever workflow task fails
+	incrementAttempt = incrementAttempt && !m.ms.IsStickyTaskQueueEnabled()
```
Comment (yiminc) [inline service/history/workflow/workflow_task_state_machine.go:559]: why this? we should increase attempt regardless sticky or not.

**Comment by wxing1292** [comment]
> I could be wrong, but I think we intentionally skip record events for continues WFT failure to avoid growing the history events indefinitely. Those are called transient events? Those events are only recorded in history when it eventually succeed or when workflow closed.

yes, that is for transient workflow task failure / timeout (normal workflow task)
this issue (PR) tries to solve weird edge case for sticky workflow task to normal workflow task transition

File: service/history/workflow/workflow_task_state_machine.go (hunk 1 above)
Comment (wxing1292) [reply service/history/workflow/workflow_task_state_machine.go:559]: sticky workflow task should not be retied to the same task queue
meaning sticky workflow task should only have attempt == 1, plus the logic is based on number of attempts to decide whether to generate new events, etc

mid / long term story: we should just throw away the existing impl and start from stretch

**Review by yiminc — APPROVED** [review]

**Pull request merged**